rewriter/Data.cc

```diff
@@ -15,30 +15,70 @@ namespace sorbet::rewriter {
 
 namespace {
 
-bool isMissingInitialize(const core::GlobalState &gs, const ast::Send *send) {
+const ast::Send *findParams(ast::ExpressionPtr *send) {
+    ast::Send *sig = ASTUtil::castSig(*send);
+    if (sig == nullptr) {
+        return nullptr;
+    }
+
+    auto *block = sig->block();
+    if (block == nullptr) {
+        return nullptr;
+    }
+
+    auto *bodyBlock = ast::cast_tree<ast::Send>(block->body);
+
+    while (bodyBlock && bodyBlock->fun != core::Names::params()) {
+        bodyBlock = ast::cast_tree<ast::Send>(bodyBlock->recv);
+    }
+
+    return bodyBlock;
+}
+
+optional<pair<ast::MethodDef *, const ast::Send *>> getInitialize(const core::GlobalState &gs, ast::Send *send) {
     if (!send->hasBlock()) {
-        return true;
+        return {};
     }
 
     auto block = send->block();
 
     if (auto *insSeq = ast::cast_tree<ast::InsSeq>(block->body)) {
-        auto methodDef = ast::cast_tree<ast::MethodDef>(insSeq->expr);
+        ast::ExpressionPtr *prevStat = nullptr;
+        for (auto &&stat : insSeq->stats) {
+            auto methodDef = ast::cast_tree<ast::MethodDef>(stat);
 
+            if (methodDef && methodDef->name == core::Names::initialize()) {
+                const ast::Send *sig = findParams(prevStat);
+                return {{methodDef, sig}};
+            }
+
+            prevStat = &stat;
+        }
+
+        // the last expression of the block is stored separately as expr
+        auto methodDef = ast::cast_tree<ast::MethodDef>(insSeq->expr);
         if (methodDef && methodDef->name == core::Names::initialize()) {
-            return false;
+            const ast::Send *sig = findParams(prevStat);
+            return {{methodDef, sig}};
         }
+    }
 
-        for (auto &&stat : insSeq->stats) {
-            methodDef = ast::cast_tree<ast::MethodDef>(stat);
+    return {};
+}
 
-            if (methodDef && methodDef->name == core::Names::initialize()) {
-                return false;
+ast::ExpressionPtr getMemberType(core::MutableContext ctx, const ast::Send *params, core::NameRef name,
+                                 core::LocOffsets loc) {
+    if (params) {
+        for (int i = 0; i < params->numKwArgs(); i++) {
+            auto key = ast::cast_tree<ast::Literal>(params->getKwKey(i))->asName();
+
+            if (key.toString(ctx) == name.toString(ctx)) {
+                return params->getKwValue(i).deepCopy();
             }
         }
     }
 
-    return true;
+    return ast::MK::Constant(loc, core::Symbols::BasicObject());
```

Comment:
I would like to be really intentional about what the expected behavior should be here.

The way it's currently implemented says that if you specify a signature, but that signature either does not have any `params` or the `params` do not mention a type for the `name` property, we treat that property's type as `BasicObject`:

- There are no tests for either of these cases.
- It's not clear that we want to use `BasicObject`. In general, I view `BasicObject` as mostly a smell these days. `T.untyped` or `T.anything` are usually better choices
- It's not clear whether we want to allow this behavior if some of the fields have been omitted. e.g. maybe the behavior should be that you can only get a typed `Data` class if the signature mentions all of the fields, instead of only some of them